internal/adapters/consul/http.go

```diff
@@ -281,42 +289,54 @@ func (f *flattenedRouteMap) flatten(route core.HTTPRoute) {
 					},
 				}}
 			}
+
+			// Add one new rule for each hostname+rule+match combination
 			for _, match := range rule.Matches {
-				found = append(found, flattenedRoute{
+				routesForHost = append(routesForHost, flattenedRoute{
 					match:    match,
 					filters:  rule.Filters,
 					services: rule.Services,
 				})
 			}
 		}
-		f.flattenedRoutesByHostname[host] = found
+
+		f.flattenedRoutesByHostname[host] = routesForHost
 	}
 }
 
+// constructRoutes consolidates all flattened route matching rules into the shortest possible
+// list of routes with one route per hostname containing all rules for that hostname.
 func (f *flattenedRouteMap) constructRoutes(gateway core.ResolvedGateway) []core.HTTPRoute {
-	coreRoutes := []core.HTTPRoute{}
-	for hostname, routes := range f.flattenedRoutesByHostname {
-		sort.SliceStable(routes, func(i, j int) bool {
-			return compareHTTPRules(routes[i].match, routes[j].match)
-		})
-		rules := []core.HTTPRouteRule{}
-		for _, match := range routes {
-			rules = append(rules, core.HTTPRouteRule{
-				Matches:  []core.HTTPMatch{match.match},
-				Filters:  match.filters,
-				Services: match.services,
-			})
-		}
-		name := gateway.ID.Service + "-" + hostsKey(hostname)
-		coreRoutes = append(coreRoutes, core.HTTPRoute{
+	var routes []core.HTTPRoute
+
+	for hostname, rules := range f.flattenedRoutesByHostname {
```

Comment:
I'm actually wondering if, given the way that this logic all works, it might make sense to change `flattenedRouteByHostname` to `rulesByHostname` or something similar. Thoughts?